src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java

```diff
@@ -173,46 +175,57 @@ public boolean executeTask(
+  public PantsExecuteTaskResult doPantsCompile(@NotNull Project project) {
+    return executeCompileTask(project, getTargetAddressesToCompile(ModuleManager.getInstance(project).getModules()), false);
   }
 
-  public Pair<Boolean, Optional<String>> executeTask(@NotNull Module[] modules) {
+  public PantsExecuteTaskResult doPantsCompile(@NotNull Module[] modules) {
     if (modules.length == 0) {
-      return Pair.create(false, Optional.empty());
+      return PantsExecuteTaskResult.emptyFailure();
     }
-    return executeTask(modules[0].getProject(), getTargetAddressesToCompile(modules), false);
+    return executeCompileTask(modules[0].getProject(), getTargetAddressesToCompile(modules), false);
   }
 
   /**
    * @param currentProject:           current project
-   * @param targetAddressesToCompile: set of target addresses to compile
-   * @param useCleanAll:              whether to use clean-all first in Pants
-   * @return whether the execution is successful, additional message along with the execution
-   * in a Pair object.
+   * @param targetAddresses:          set of target addresses given to pants executable (e.g. "::")
+   * @param tasks:                    set of tasks given to pants executable (e.g. "lint")
+   * @param opTitle:                  simple title describing what this invocation is doing (e.g. "Compile")
+   * @return whether the execution is successful and an optional message
+   * describing the result as a PantsExecuteTaskResult object
    */
-  public Pair<Boolean, Optional<String>> executeTask(Project currentProject, Set<String> targetAddressesToCompile, boolean useCleanAll) {
-    // If project has not changed since last Compile, return immediately.
-    if (!FileChangeTracker.shouldRecompileThenReset(currentProject, targetAddressesToCompile)) {
-      PantsExternalMetricsListenerManager.getInstance().logIsPantsNoopCompile(true);
-      Notification start = new Notification(PantsConstants.PANTS, "Pants Compile", "Already up to date.\n", NotificationType.INFORMATION);
-      Notifications.Bus.notify(start);
-      return Pair.create(true, Optional.of(PantsConstants.NOOP_COMPILE));
-    }
-
+  public PantsExecuteTaskResult invokePants(@NotNull Project currentProject,
+                                            @NotNull Set<String> targetAddresses,
+                                            @NotNull List<String> tasks,
+                                            @NotNull String opTitle) {
+    // TODO: assert that opTitle is one word?
     prepareIDE(currentProject);
-    if (targetAddressesToCompile.isEmpty()) {
+
+    List<String> unrecognizedTasks = tasks.stream()
+      .filter(task -> !PantsConstants.SUPPORTED_TASKS.contains(task))
+      .collect(Collectors.toList());
+    if (!unrecognizedTasks.isEmpty()) {
+      String unrecognizedMessage = String.format("Unrecognized tasks specified in Pants invocation: [%s].\n",
```

Comment:
this check is probably unnecessary. if the tasks aren't supported, pants will give the corresponding error messages.

Reply:
Thanks, removed.